src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+        public FileGrainStorage(string storageName, string rootDirectory, IGrainFactory grainFactory, ITypeResolver typeResolver)
+        {
+            _storageName = storageName;
+            _rootDirectory = rootDirectory;
+            _grainFactory = grainFactory;
+            _typeResolver = typeResolver;
+        }
+
+        public Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+  
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
+```
+
+We create a constructor containing two fields, `storageName` to specify which grains should write using this storage `[StorageProvider(ProviderName = "File")]` and `directory` which would be the directory where the grain states will be saved.
+
+`IGrainFactory`, `ITypeResolver` and `JsonSerializerSettings` will be used in the next section where we will initilize the storage.
+
+## Initializing the storage
+
+To initialize the storage, we register an `Init` function on the `ApplicationServices` lifecycle.
+
+```csharp
+public void Participate(ISiloLifecycle lifecycle)
+{
+    lifecycle.Subscribe(OptionFormattingUtilities.Name<FileGrainStorage>(_storageName), ServiceLifecycleStage.ApplicationServices, Init);
+}
+```
+
+The `Init` function is used to set the `_jsonSettings` which will be used to configure the `Json` serializer. At the same time we create the folder to store the grains states if it does not exist yet.
+
+```csharp
+private Task Init(CancellationToken ct)
+{
+    // Settings could be made configurable from Options.
+    _jsonSettings = OrleansJsonSerializer.UpdateSerializerSettings(OrleansJsonSerializer.GetDefaultSerializerSettings(_typeResolver, _grainFactory), false, false, null);
+
+    var directory = new System.IO.DirectoryInfo(_rootDirectory);
+    if (!directory.Exists)
+        directory.Create();
+
+    return Task.CompletedTask;
+}
+```
+
+## Reading State
+
+To read the state, to ensure uniqueness, we construct the filename using the grain key and the grain type.
+
+```csharp
+public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (!fileInfo.Exists)
+        return;
```

Comment:
If no state is found, probably want to set state to default.  Something like:

grainState.State = Activator.CreateInstance(grainState.State.GetType())

Reply:
Nice, if state not found, I added `grainState.State = Activator.CreateInstance(grainState.State.GetType());` to set the state to a `default` instantiation.